### DISCUSSION TITLE
Detect BLE connection interval in sketch?

### DISCUSSION THREAD

**Discussion by amowry** [adafruit/Adafruit_nRF52_Arduino#781] [Q&A] [answered]
Is there a way to detect the current BLE connection interval in a sketch? This would be useful for optimizing the data that is sent over BLE.

**Comment by hathach** [comment]
check the code and/or example, there is plenty reference there

**Reply by amowry** [reply]
Thank you-- I've read the examples but I haven't seen a way of detecting the negotiated connection interval, only the requested interval. I wasn't sure if there's a method that I haven't seen (?).

**Reply by santatester** [reply, marked as answer]
I couldn't find it in any of the examples, but there is a 'getConnectionInterval' function that will give you the result you're looking for.

Here's the peripheral 'bleuart' example's connect_callback modified with the connection interval function:


```cpp
void connect_callback(uint16_t conn_handle)
{
  // Get the reference to current connection
  BLEConnection* connection = Bluefruit.Connection(conn_handle);

  char central_name[32] = { 0 };
  uint16_t connIntvl = 0;

  connection->getPeerName(central_name, sizeof(central_name));

// Get the current connection agreed upon connection interval
  connIntvl = connection->getConnectionInterval();

  Serial.print("Connected to ");
  Serial.println(central_name);

  Serial.print("At connection interval ");
  Serial.println(connIntvl);
}
```

Hope that helps...

**Reply by amowry** [reply]
That worked perfectly. Thanks so much for your help!

**Comment by amowry** [comment]
I've noticed that this sometimes returns a different connection interval than is reported by the "Full debug" output in the IDE. In the example below, the debug output reports a 7.5 ms connection interval (which I believe is correct) while connection->getConnectionInterval() returns an interval of 6 (which is 3.75 ms). This is shown in the last line below (sent from the connect callback). Any idea why this happens?

09:51:40.126 -> [BLE   ] BLE_GAP_EVT_CONNECTED : Conn Handle = 0
09:51:40.126 -> [GAP   ] MAC = 20:98:66:65:83:6A, Type = 0, Resolved = 0
09:51:40.126 -> [GAP   ] Conn Interval = 7.50 ms, Latency = 0, Supervisor Timeout = 30000 ms
09:51:40.126 -> [BLE   ] BLE_GATTC_EVT_CHAR_VAL_BY_UUID_READ_RSP : Conn Handle = 0
09:51:40.126 -> connIntvl = 6
